actors/builtin/miner/miner_test.go

```diff
@@ -2052,15 +2052,178 @@ func TestWithdrawBalance(t *testing.T) {
 
 func TestChangePeerID(t *testing.T) {
 	periodOffset := abi.ChainEpoch(100)
+	actor := newHarness(t, periodOffset)
+	builder := builderForHarness(actor).
+		WithBalance(bigBalance, big.Zero())
 
 	t.Run("successfully change peer id", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt)
+
+		newPID := tutil.MakePID("test-change-peer-id")
+		actor.changePeerID(rt, newPID)
+	})
+}
+
+func TestCompactPartitions(t *testing.T) {
+	periodOffset := abi.ChainEpoch(100)
+	actor := newHarness(t, periodOffset)
+	builder := builderForHarness(actor).
+		WithBalance(bigBalance, big.Zero())
+
+	assertSectorExists := func(store adt.Store, st *miner.State, sectorNum abi.SectorNumber, expectPart uint64, expectDeadline uint64) {
+		deadline, pid, err := st.FindSector(store, sectorNum)
+		require.NoError(t, err)
+		require.EqualValues(t, expectPart, pid)
+		require.EqualValues(t, expectDeadline, deadline)
+	}
+
+	t.Run("compacting a partition with both live and dead sectors removes the dead sectors but retains the live sectors", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt)
+
+		rt.SetEpoch(200)
+		// create 4 sectors in partition 0
+		info := actor.commitAndProveSectors(rt, 4, defaultSectorExpiration, [][]abi.DealID{{10}, {20}, {30}, {40}})
+		sector1 := info[0].SectorNumber
+		sector2 := info[1].SectorNumber
+		sector3 := info[2].SectorNumber
+		sector4 := info[3].SectorNumber
+
+		// terminate sector1
+		rt.SetEpoch(rt.Epoch() + 100)
+		actor.addLockedFunds(rt, big.Mul(big.NewInt(1e18), big.NewInt(20000)))
+		tsector := info[0]
+		sectorSize, err := tsector.SealProof.SectorSize()
+		require.NoError(t, err)
+		sectorPower := miner.QAPowerForSector(sectorSize, tsector)
+		dayReward := miner.ExpectedRewardForPower(actor.epochRewardSmooth, actor.epochQAPowerSmooth, sectorPower, builtin.EpochsInDay)
+		twentyDayReward := miner.ExpectedRewardForPower(actor.epochRewardSmooth, actor.epochQAPowerSmooth, sectorPower, miner.InitialPledgeProjectionPeriod)
+		sectorAge := rt.Epoch() - tsector.Activation
+		expectedFee := miner.PledgePenaltyForTermination(dayReward, sectorAge, twentyDayReward, actor.epochQAPowerSmooth, sectorPower, actor.epochRewardSmooth, big.Zero(), 0)
+
+		sectors := bitfield.NewFromSet([]uint64{uint64(sector1)})
+		actor.terminateSectors(rt, sectors, expectedFee)
+
+		// compacting partition will remove sector1 but retain sector 2, 3 and 4.
+		partId := uint64(0)
+		deadlineId := uint64(0)
+		partitions := bitfield.NewFromSet([]uint64{partId})
+		actor.compactPartitions(rt, deadlineId, partitions)
+
+		st := getState(rt)
+		assertSectorExists(rt.AdtStore(), st, sector2, partId, deadlineId)
+		assertSectorExists(rt.AdtStore(), st, sector3, partId, deadlineId)
+		assertSectorExists(rt.AdtStore(), st, sector4, partId, deadlineId)
+
+		// TODO Why is this working ?
+		assertSectorExists(rt.AdtStore(), st, sector1, partId, deadlineId)
```

Comment:
@Stebalien I am not sure what's going wrong here.

This sector was terminated above and thus should not exist after compacting the partition. Please can you review the test ?

Reply:
Because the code wasn't saving the state... I've pushed a fix.